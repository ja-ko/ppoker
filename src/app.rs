use std::error;
use std::time::{Duration, Instant};
use log::{debug, error, info};
use notify_rust::{Notification, Timeout};
use crate::web::client::PokerClient;
use crate::config::Config;
use crate::models::{GamePhase, LogEntry, LogLevel, LogSource, Player, Room, Vote, VoteData};

pub type AppResult<T> = std::result::Result<T, Box<dyn error::Error>>;

pub struct App {
    pub running: bool,
    pub vote: Option<VoteData>,
    pub name: String,

    pub room: Room,
    pub client: PokerClient,
    pub log: Vec<LogEntry>,

    pub round_number: u32,

    pub config: Config,

    pub has_focus: bool,
    notify_vote_at: Option<Instant>,
    is_notified: bool,
}

impl App {
    pub fn new(config: Config) -> AppResult<Self> {
        let (client, room, log) = PokerClient::new(&config)?;

        let mut result = Self {
            running: true,
            vote: None,
            name: config.name.clone(),
            room,
            client,
            log: vec![],
            round_number: 0,
            config,
            has_focus: true,
            notify_vote_at: None,
            is_notified: false,
        };
        result.update_server_log(log);

        Ok(result)
    }

    pub fn tick(&mut self) {
        if let Some(notify_at) = &self.notify_vote_at {
            if *notify_at < Instant::now() && !self.is_notified {
                if self.has_focus {
                    info!("Skipping notification because user has application focused.")
                } else {
                    info!("Notifying user of missing vote.");
                    if let Err(e) = Notification::new()
                        .summary("Planning Poker")
                        .body("Your vote is the last one missing.")
                        .timeout(Timeout::Milliseconds(10000))
                        .show() {
                        error!("Failed to send notification: {}", e);
                    }
                }
                self.is_notified = true;
                self.notify_vote_at = None;
            }
        }  
    }

    #[inline]
    fn deck_has_value(&self, vote: u8) -> bool {
        self.room.deck.iter().find(|item| vote.to_string().eq_ignore_ascii_case(item)).is_some()
    }

    #[inline]
    fn is_my_vote_last_missing(&self) -> bool {
        self.room.players.len() > 1
            && self.room.players.iter().filter(|p| p.vote == Vote::Missing).count() == 1
            && self.vote.is_none()
            && self.room.phase == GamePhase::Playing
    }

    pub fn merge_update(&mut self, update: Room) {
        debug!("room update: {:?}", update);
        if update.phase == GamePhase::Playing && self.room.phase != GamePhase::Playing {
            self.vote = None;
            self.round_number += 1;
            self.is_notified = false;
            self.notify_vote_at = None;
        }
        self.room = update;
        if self.is_my_vote_last_missing() {
            if !self.is_notified && self.notify_vote_at == None {
                self.log_message(LogLevel::Info, "Your vote is the last one missing.".to_string());
                self.notify_vote_at = Some(Instant::now() + Duration::from_secs(15));
            }
        } else {
            self.notify_vote_at = None;
        }
    }

    pub fn vote(&mut self, data: &str) -> AppResult<()> {
        let vote = data.parse::<u8>();
        if vote.is_ok() {
            let vote = vote.unwrap();
            if self.deck_has_value(vote) {
                let vote = VoteData::Number(vote);
                self.client.vote(Some(format!("{}", &vote).as_str()))?;
                self.vote = Some(vote);
            } else {
                self.log_message(LogLevel::Error, format!("Card is not in the deck: {}", data));
            }
        } else {
            self.vote = None;
            self.client.vote(None)?;
            if data != "-" {
                self.log_message(LogLevel::Error, format!("Unable to parse card: {}", data));
            }
        }
        Ok(())
    }

    pub fn rename(&mut self, data: String) -> AppResult<()> {
        self.name = data;
        self.client.change_name(self.name.as_str())?;

        Ok(())
    }

    pub fn reveal(&mut self) -> AppResult<()> {
        self.client.reveal()
    }

    pub fn chat(&mut self, message: String) -> AppResult<()> {
        self.client.chat(message.as_str())
    }

    pub fn restart(&mut self) -> AppResult<()> {
        self.vote = None;
        self.client.reset()
    }

    pub fn update(&mut self) -> AppResult<()> {
        let (room_updates, log_updates) = self.client.get_updates()?;
        // TODO: reconnect?

        for update in room_updates {
            self.merge_update(update);
        }

        self.update_server_log(log_updates);

        Ok(())
    }

    fn update_server_log(&mut self, log_updates: Vec<LogEntry>) {
        for log in log_updates {
            if self.log.iter().find(|l| l.server_index == log.server_index).is_none() {
                self.log.push(log);
            }
        }
    }

    pub fn log_message(&mut self, level: LogLevel, message: String) {
        self.log.push(LogEntry {
            timestamp: Instant::now(),
            level,
            message,
            source: LogSource::Client,
            server_index: None,
        })
    }
}