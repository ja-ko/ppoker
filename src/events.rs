use std::sync::mpsc;
use std::thread;
use std::time::{Duration, Instant};
use crossterm::event;
use crossterm::event::{KeyEvent, KeyEventKind, MouseEvent, Event as CrosstermEvent};
use crate::app::AppResult;

#[derive(Clone, Copy, Debug)]
pub enum FocusChange {
    Gained,
    Lost
}

#[derive(Clone, Copy, Debug)]
pub enum Event {
    Tick,
    Key(KeyEvent),
    Mouse(MouseEvent),
    Resize(u16, u16),
    Focus(FocusChange)
}

#[derive(Debug)]
pub struct EventHandler {
    receiver: mpsc::Receiver<Event>,
    shutdown: mpsc::Sender<()>,
    handler: thread::JoinHandle<()>,
}

impl EventHandler {
    pub fn new(tick_rate: u64) -> Self {
        let tick_rate = Duration::from_millis(tick_rate);
        let (sender, receiver) = mpsc::channel();
        let (shutdown, shutdown_recv) = mpsc::channel();
        let handler = {
            let sender = sender.clone();
            thread::spawn(move || {
                let mut last_tick = Instant::now();
                loop {
                    if shutdown_recv.try_recv().is_ok() {
                        break;
                    }
                    
                    let timeout = tick_rate
                        .checked_sub(last_tick.elapsed())
                        .unwrap_or(tick_rate);

                    if event::poll(timeout).expect("failed to poll new events") {
                        match event::read().expect("unable to read event") {
                            CrosstermEvent::Key(e) => {
                                if e.kind == KeyEventKind::Press {
                                    sender.send(Event::Key(e))
                                } else {
                                    Ok(())
                                }
                            }
                            CrosstermEvent::Mouse(e) => sender.send(Event::Mouse(e)),
                            CrosstermEvent::Resize(w, h) => sender.send(Event::Resize(w, h)),
                            CrosstermEvent::FocusGained => sender.send(Event::Focus(FocusChange::Gained)),
                            CrosstermEvent::FocusLost => sender.send(Event::Focus(FocusChange::Lost)),
                            CrosstermEvent::Paste(_) => unimplemented!(),
                        }
                            .expect("failed to send terminal event")
                    }

                    if last_tick.elapsed() >= tick_rate {
                        sender.send(Event::Tick).expect("failed to send tick event");
                        last_tick = Instant::now();
                    }
                    
                }
            })
        };
        Self {
            receiver,
            handler,
            shutdown,
        }
    }

    pub fn next(&self) -> AppResult<Event> {
        Ok(self.receiver.recv()?)
    }

    pub fn shutdown(self) {
        self.shutdown.send(()).expect("Unable to signal event thread to shutdown");
        self.handler.join().expect("Unable to join event thread.");
    }
}

