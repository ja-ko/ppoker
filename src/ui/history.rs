use crossterm::event::{KeyCode, KeyEvent};
use ratatui::layout::{Constraint, Direction, Layout};
use ratatui::prelude::*;
use ratatui::widgets::{Cell, Row, Table, TableState};
use ratatui::Frame;

use crate::app::{App, AppResult, HistoryEntry};
use crate::models::GamePhase;
use crate::ui::voting::{format_vote, render_overview, render_own_vote};
use crate::ui::{
    colored_box_style, footer_entries, format_duration, render_box, render_box_colored, Page,
    UIAction, UiPage,
};

pub struct HistoryPage {
    history_state: TableState,
}

impl HistoryPage {
    pub fn new() -> Self {
        Self {
            history_state: TableState::default(),
        }
    }
}

impl Page for HistoryPage {
    fn render(&mut self, app: &mut App, frame: &mut Frame) {
        if self.history_state.selected().is_none() && app.history.len() > 0 {
            self.history_state.select(Some(0));
        }

        let [header, body, footer] = Layout::default()
            .direction(Direction::Vertical)
            .constraints([
                Constraint::Length(3),
                Constraint::Fill(1),
                Constraint::Length(3),
            ])
            .areas(frame.area());

        render_overview(app, header, frame);
        self.render_main(app, body, frame);
        self.render_footer(app, footer, frame);
    }

    fn input(&mut self, _app: &mut App, event: KeyEvent) -> AppResult<UIAction> {
        return Ok(match event.code {
            KeyCode::Esc | KeyCode::Char('q') => UIAction::Quit,
            KeyCode::Char(c) if c == 'v' || c == '-' || c == 'h' || c.is_ascii_digit() => {
                UIAction::ChangeView(UiPage::Voting)
            }
            KeyCode::Down => {
                if let Some(s) = self.history_state.selected() {
                    let mut new_index = s.saturating_add(1);
                    if new_index >= _app.history.len() {
                        new_index = _app.history.len().saturating_sub(1);
                    }
                    self.history_state.select(Some(new_index));
                }
                UIAction::Continue
            }
            KeyCode::Up => {
                if let Some(s) = self.history_state.selected() {
                    self.history_state.select(Some(s.saturating_sub(1)));
                }
                UIAction::Continue
            }
            _ => UIAction::Continue,
        });
    }
}

impl HistoryPage {
    fn render_main(&mut self, app: &mut App, rect: Rect, frame: &mut Frame) {
        let [history, detail] =
            Layout::horizontal([Constraint::Percentage(40), Constraint::Percentage(60)])
                .areas(rect);

        let [vote_summary, players] =
            Layout::vertical([Constraint::Length(9), Constraint::Fill(1)]).areas(detail);

        let current_entry = self.history_state.selected().map(|idx| &app.history[idx]);

        if let Some(current_entry) = current_entry {
            render_own_vote(
                &current_entry.votes,
                current_entry.average,
                GamePhase::Revealed,
                &current_entry.own_vote,
                &current_entry.deck,
                vote_summary,
                frame,
            );

            render_player_list(&current_entry, players, frame);
        }
        self.render_history(app, history, frame);
    }

    fn render_footer(&mut self, app: &mut App, rect: Rect, frame: &mut Frame) {
        let entries = vec!["Vote", "↑", "↓", "Quit"];
        let mut footer = footer_entries(entries);
        if app.has_updates {
            footer = footer.style(Style::new().yellow());
        }
        frame.render_widget(footer, rect);
    }

    fn render_history(&mut self, app: &mut App, rect: Rect, frame: &mut Frame) {
        let inner = render_box("History", rect, frame);

        let rows: Vec<Row> = app
            .history
            .iter()
            .map(|entry| {
                Row::new(vec![
                    Cell::from(Span::raw(entry.round_number.to_string())),
                    Cell::from(Span::raw(format!("{:.1}", entry.average))),
                    Cell::from(Span::raw(format_duration(&entry.length))),
                ])
            })
            .collect();

        let table = Table::new(
            rows,
            [
                Constraint::Length(5),
                Constraint::Length(8),
                Constraint::Fill(1),
            ],
        )
        .column_spacing(4)
        .header(
            Row::new(vec!["Round", "Average", "Duration"])
                .style(Style::new().bold())
                .bottom_margin(1),
        )
        .highlight_symbol("> ")
        .row_highlight_style(Style::new().on_white().black());

        frame.render_stateful_widget(table, inner, &mut self.history_state);
    }
}

fn render_player_list(entry: &HistoryEntry, rect: Rect, frame: &mut Frame) {
    let inner = render_box_colored(
        "Players",
        colored_box_style(GamePhase::Revealed),
        rect,
        frame,
    );
    let mut longest_name = 0;
    let mut players = entry.votes.clone();
    players.sort();
    let rows: Vec<Row> = players
        .iter()
        .map(|p| {
            if p.name.len() > longest_name {
                longest_name = p.name.len();
            }
            Row::new(vec![
                Cell::from(Span::raw(p.name.as_str())),
                Cell::from(format_vote(&p.vote, &entry.own_vote)),
            ])
        })
        .collect();

    let table = Table::new(
        rows,
        [Constraint::Length(longest_name as u16), Constraint::Fill(1)],
    )
    .column_spacing(4)
    .header(
        Row::new(vec!["Name", "Vote"])
            .style(Style::new().bold())
            .bottom_margin(1),
    );

    frame.render_widget(table, inner);
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::app::tests::create_test_app;
    use crate::ui::tests::{send_input, tick};
    use crate::web::client::tests::LocalMockPokerClient;
    use crate::web::client::PokerClient;
    use insta::assert_snapshot;
    use ratatui::backend::TestBackend;

    #[test]
    fn test_render_page() {
        let mut page = HistoryPage::new();
        let mut app = create_test_app(Box::new(LocalMockPokerClient::new("test")));
        let mut terminal = Terminal::new(TestBackend::new(80, 20)).unwrap();
        tick(&mut terminal, &mut page, &mut app);

        assert_snapshot!("Empty history page", terminal.backend());
    }

    #[test]
    fn test_render_page_with_history() {
        let mut page = HistoryPage::new();
        let mut client = LocalMockPokerClient::new("Alice");

        // Add other players
        let bob_id = client.add_user("Bob");
        let charlie_id = client.add_user("Charlie");

        // First round: Everyone votes numbers
        client.vote(Some("5")).unwrap();
        client.user_vote(&bob_id, Some("3"));
        client.user_vote(&charlie_id, Some("8"));
        client.reveal().unwrap();
        client.reset().unwrap();

        // Second round: Mix of numbers and special votes
        client.vote(Some("13")).unwrap();
        client.user_vote(&bob_id, Some("?"));
        client.user_vote(&charlie_id, Some("8"));
        client.reveal().unwrap();
        client.reset().unwrap();

        // Third round: Some abstentions
        client.vote(Some("3")).unwrap();
        client.user_vote(&bob_id, Some("5"));
        // Charlie doesn't vote
        client.reveal().unwrap();
        client.reset().unwrap();

        let mut app = create_test_app(Box::new(client));

        // Render and snapshot the history page
        let mut terminal = Terminal::new(TestBackend::new(90, 30)).unwrap();
        tick(&mut terminal, &mut page, &mut app);

        assert_snapshot!("History page with multiple rounds", terminal.backend());

        send_input(KeyCode::Down, &mut terminal, &mut page, &mut app);
        assert_snapshot!("History page after pressing down", terminal.backend());
    }
}
