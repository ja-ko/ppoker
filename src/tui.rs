use std::collections::HashMap;
use std::{io, panic};

use crossterm::event::{
    DisableBracketedPaste, DisableFocusChange, EnableBracketedPaste, EnableFocusChange, KeyEvent,
};
use crossterm::terminal;
use crossterm::terminal::{EnterAlternateScreen, LeaveAlternateScreen};
use log::debug;
use ratatui::prelude::*;

use crate::app::{App, AppResult};
use crate::events::{Event, EventHandler, FocusChange};
use crate::ui::HistoryPage;
use crate::ui::LogPage;
use crate::ui::VotingPage;
use crate::ui::{Page, UIAction, UiPage};

pub struct Tui<B: Backend> {
    terminal: Terminal<B>,
    pub events: EventHandler,
    pub current_page: UiPage,
    pages: HashMap<UiPage, Box<dyn Page>>,
}

impl<B: Backend> Tui<B> {
    pub fn new(terminal: Terminal<B>, events: EventHandler) -> Self {
        let mut pages: HashMap<UiPage, Box<dyn Page>> = HashMap::new();
        enum_iterator::all::<UiPage>().for_each(|page| match page {
            UiPage::Voting => {
                pages.insert(page, Box::new(VotingPage::new()));
            }
            UiPage::Log => {
                pages.insert(page, Box::new(LogPage::new()));
            }
            UiPage::History => {
                pages.insert(page, Box::new(HistoryPage::new()));
            }
        });
        Self {
            terminal,
            events,
            current_page: UiPage::Voting,
            pages,
        }
    }
    pub fn init(&mut self) -> AppResult<()> {
        terminal::enable_raw_mode()?;
        crossterm::execute!(
            io::stderr(),
            EnterAlternateScreen,
            EnableFocusChange,
            EnableBracketedPaste
        )?;

        let panic_hook = panic::take_hook();
        panic::set_hook(Box::new(move |panic| {
            tui_logger::move_events();
            Self::reset().expect("failed to reset the terminal");
            panic_hook(panic);
        }));

        self.terminal.hide_cursor()?;
        self.terminal.clear()?;
        Ok(())
    }

    pub fn draw(&mut self, app: &mut App) -> AppResult<()> {
        let page = self.pages.get_mut(&self.current_page).unwrap();
        self.terminal.draw(|frame| page.render(app, frame))?;
        Ok(())
    }

    fn reset() -> AppResult<()> {
        terminal::disable_raw_mode()?;
        crossterm::execute!(
            io::stderr(),
            LeaveAlternateScreen,
            DisableFocusChange,
            DisableBracketedPaste
        )?;
        Ok(())
    }

    pub fn exit(mut self) -> AppResult<()> {
        self.events.shutdown();
        Self::reset()?;
        self.terminal.show_cursor()?;
        Ok(())
    }

    pub fn handle_events(&mut self, app: &mut App) -> AppResult<()> {
        match self.events.next()? {
            Event::Tick => app.tick()?,
            Event::Key(event) => self.handle_key(event, app)?,
            Event::Mouse(_) => {}
            Event::Resize(_, _) => {}
            Event::Focus(change) => {
                debug!("Focus change: {:?}", change);
                match change {
                    FocusChange::Gained => {
                        app.has_focus = true;
                    }
                    FocusChange::Lost => {
                        app.has_focus = false;
                    }
                }
            }
            Event::Paste(text) => self
                .pages
                .get_mut(&self.current_page)
                .unwrap()
                .pasted(app, text),
        }
        Ok(())
    }

    fn handle_key(&mut self, key_event: KeyEvent, app: &mut App) -> AppResult<()> {
        let page = self.pages.get_mut(&self.current_page).unwrap();
        let action = page.input(app, key_event)?;
        match action {
            UIAction::Continue => {}
            UIAction::ChangeView(page) => self.current_page = page,
            UIAction::Quit => {
                app.running = false;
            }
        }
        Ok(())
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::app::tests::create_test_app;
    use crate::web::client::tests::LocalMockPokerClient;
    use crossterm::event::{KeyCode, KeyModifiers};
    #[cfg(test)]
    use insta::assert_snapshot;
    use ratatui::backend::TestBackend;

    fn create_test_terminal() -> Terminal<TestBackend> {
        Terminal::new(TestBackend::new(80, 24)).unwrap()
    }

    fn create_test_tui() -> (Tui<TestBackend>, App) {
        let terminal = create_test_terminal();
        let events = EventHandler::new(100);
        let tui = Tui::new(terminal, events);
        let app = create_test_app(Box::new(LocalMockPokerClient::new("test")));
        (tui, app)
    }

    #[test]
    fn test_page_switching() {
        let (mut tui, mut app) = create_test_tui();

        // Initially should be on voting page
        assert_eq!(tui.current_page, UiPage::Voting);
        tui.draw(&mut app).unwrap();
        assert_snapshot!("initial_voting_page", tui.terminal.backend());

        // Switch to history page with 'h'
        tui.handle_key(
            KeyEvent::new(KeyCode::Char('h'), KeyModifiers::empty()),
            &mut app,
        )
        .unwrap();
        assert_eq!(tui.current_page, UiPage::History);
        tui.draw(&mut app).unwrap();
        assert_snapshot!("switched_to_history", tui.terminal.backend());

        // Switch back to voting page with 'v'
        // This is done because you can only switch to the log page from the voting page
        tui.handle_key(
            KeyEvent::new(KeyCode::Char('v'), KeyModifiers::empty()),
            &mut app,
        )
        .unwrap();
        assert_eq!(tui.current_page, UiPage::Voting);
        tui.draw(&mut app).unwrap();

        // Switch to log page with 'l'
        tui.handle_key(
            KeyEvent::new(KeyCode::Char('l'), KeyModifiers::empty()),
            &mut app,
        )
        .unwrap();
        assert_eq!(tui.current_page, UiPage::Log);
        tui.draw(&mut app).unwrap();
        assert_snapshot!("switched_to_log", tui.terminal.backend());

        // Switch back to voting page with 'l'
        tui.handle_key(
            KeyEvent::new(KeyCode::Char('l'), KeyModifiers::empty()),
            &mut app,
        )
        .unwrap();
        assert_eq!(tui.current_page, UiPage::Voting);
        tui.draw(&mut app).unwrap();
        assert_snapshot!("back_to_voting", tui.terminal.backend());
    }
}
