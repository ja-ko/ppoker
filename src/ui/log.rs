use crossterm::event::{KeyCode, KeyEvent};
use log::LevelFilter;
use ratatui::Frame;
use ratatui::layout::{Constraint, Layout};
use ratatui::prelude::*;
use ratatui::style::Style;
use ratatui::widgets::{Block, BorderType, Paragraph, Wrap};
use tui_logger::{TuiLoggerLevelOutput, TuiLoggerSmartWidget, TuiWidgetEvent, TuiWidgetState};

use crate::app::{App, AppResult};
use crate::tui::UiPage;
use crate::ui::{Page, UIAction};

pub struct LogPage {
    state: TuiWidgetState,
}

impl LogPage {
    pub fn new() -> Self {
        Self {
            state: TuiWidgetState::default()
                .set_level_for_target("tungstenite::client", LevelFilter::Warn)
                .set_level_for_target("tungstenite::handshake::client", LevelFilter::Warn)
                .set_level_for_target("ppoker::web::ws", LevelFilter::Info)
        }
    }
}

impl Page for LogPage {
    fn render(&mut self, _app: &mut App, frame: &mut Frame) {
        let mut helptexts: Vec<Span> = vec![];
        helptexts.append(&mut help_spans("h", "Toggle target selector"));
        helptexts.append(&mut help_spans("f", "Toggle focus"));
        helptexts.append(&mut help_spans("UP/DOWN", "Navigate"));
        helptexts.append(&mut help_spans("LEFT/RIGHT", "Reduce/increase level"));
        helptexts.append(&mut help_spans("PAGEUP/PAGEDOWN", "Enter Page mode, scroll up/down"));
        helptexts.append(&mut help_spans("ESCAPE", "Exit page mode"));
        helptexts.append(&mut help_spans("SPACE", "Toggle hiding disabled targets"));
        helptexts.append(&mut help_spans("l", "Leave log view"));
        helptexts.append(&mut help_spans("q", "Quit application"));
        helptexts.pop();

        let help_paragraph = Paragraph::new(Line::from(helptexts))
            .wrap(Wrap { trim: true });

        let help_lines = help_paragraph.line_count(frame.size().width.saturating_sub(2)) as u16;

        let [log, help] = Layout::default()
            .direction(Direction::Vertical)
            .constraints([
                Constraint::Fill(1),
                Constraint::Length(help_lines + 2)
            ])
            .areas(frame.size());

        let widget = TuiLoggerSmartWidget::default()
            .style_error(Style::default().red())
            .style_debug(Style::default().cyan())
            .style_warn(Style::default().light_yellow())
            .style_info(Style::default().white())
            .output_separator('|')
            .output_timestamp(Some("%H:%M:%S".to_string()))
            .output_level(Some(TuiLoggerLevelOutput::Long))
            .output_target(true)
            .output_file(false)
            .output_line(false)
            .state(&self.state);

        frame.render_widget(widget, log);

        let block = Block::bordered()
            .title("Help")
            .title_alignment(Alignment::Left)
            .border_type(BorderType::Rounded);
        let help_inner = block.inner(help);

        frame.render_widget(block, help);

        frame.render_widget(help_paragraph, help_inner);
    }

    fn input(&mut self, _app: &mut App, event: KeyEvent) -> AppResult<UIAction> {
        match event.code.into() {
            KeyCode::Char('q') => return Ok(UIAction::Quit),
            KeyCode::Char(' ') => self.state.transition(TuiWidgetEvent::SpaceKey),
            KeyCode::Esc => self.state.transition(TuiWidgetEvent::EscapeKey),
            KeyCode::PageUp => self.state.transition(TuiWidgetEvent::PrevPageKey),
            KeyCode::PageDown => self.state.transition(TuiWidgetEvent::NextPageKey),
            KeyCode::Up => self.state.transition(TuiWidgetEvent::UpKey),
            KeyCode::Down => self.state.transition(TuiWidgetEvent::DownKey),
            KeyCode::Left => self.state.transition(TuiWidgetEvent::LeftKey),
            KeyCode::Right => self.state.transition(TuiWidgetEvent::RightKey),
            KeyCode::Char('h') => self.state.transition(TuiWidgetEvent::HideKey),
            KeyCode::Char('f') => self.state.transition(TuiWidgetEvent::FocusKey),
            KeyCode::Char('l') => return Ok(UIAction::ChangeView(UiPage::Voting)),
            _ => {}
        }
        return Ok(UIAction::Continue);
    }
}

fn help_spans<'a>(key: &'a str, description: &'a str) -> Vec<Span<'a>> {
    vec![
        Span::raw(key).style(Style::new().bold()),
        Span::raw(" - ").style(Style::new().gray()),
        Span::raw(description),
        Span::raw(" | ").style(Style::new().gray()),
    ]
}