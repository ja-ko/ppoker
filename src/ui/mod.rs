use crossterm::event::KeyEvent;
use ratatui::Frame;
use crate::app::{App, AppResult};
use crate::tui::UiPage;

pub(crate) mod voting;
pub(crate) mod log;


pub enum UIAction{
    Continue,
    ChangeView(UiPage),
    Quit,
}
pub trait Page {
    fn render(&mut self, app: &mut App, frame: &mut Frame);
    fn input(&mut self, app: &mut App, event: KeyEvent) -> AppResult<UIAction>;
}