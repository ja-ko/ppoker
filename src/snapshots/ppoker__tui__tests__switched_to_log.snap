---
source: src/tui.rs
expression: tui.terminal.backend()
---
"┌Tui Target Selector─────────────────┐┌Tui Log [log=0.0/s]─────────────────────┐"
"│EWI  :ppoker::web::ws               ││                                        │"
"│EWI  :tungstenite::client           ││                                        │"
"│EWI  :tungstenite::handshake::client││                                        │"
"│                                    ││                                        │"
"│                                    ││                                        │"
"│                                    ││                                        │"
"│                                    ││                                        │"
"│                                    ││                                        │"
"│                                    ││                                        │"
"│                                    ││                                        │"
"│                                    ││                                        │"
"│                                    ││                                        │"
"│                                    ││                                        │"
"│                                    ││                                        │"
"│                                    ││                                        │"
"│                                    ││                                        │"
"└────────────────────────────────────┘└────────────────────────────────────────┘"
"╭Help──────────────────────────────────────────────────────────────────────────╮"
"│h - Toggle target selector | f - Toggle focus | UP/DOWN - Navigate |          │"
"│LEFT/RIGHT - Reduce/increase level | PAGEUP/PAGEDOWN - Enter Page mode, scroll│"
"│up/down | ESCAPE - Exit page mode | SPACE - Toggle hiding disabled targets | l│"
"│- Leave log view | q - Quit application                                       │"
"╰──────────────────────────────────────────────────────────────────────────────╯"
